### ISSUE TITLE
[BUG] Using Code Reformat command with Parinfer (smart-mode) breaks code

### ISSUE THREAD

**Issue by Severed-Infinity** [cursive-ide/cursive#1959]
This is my second time writing this, github keeps crashing/breaking/not loading in safari, just incase it somehow shows up twice.

Using the latest version of cursive (the most recent EAP build) on a mac (just incase it might matter) and I've an issue when using the Code Reformat command '⌥⌘L' with Parinfer (smart-mode) turned on. Using this piece of code to show the issue that happens along with supplied screenshots.

```
(defn resource-routes []
    (yada/listener
        ["/" (yada/handler (new File "app.html"))]
        {:port 5000}))
```

When using the Code Reformat command it cause the arguments to `yada/listener` to be kicked out of the function with indentation remaining the same as before (not moved from the location) but with the closing brackets moving up next to `listener` like shown here.

```
(defn resource-routes []
    (yada/listener)
        ["/" (yada/handler (new File "app.html"))]
        {:port 5000})
```
And running it again, or as I just found out while rewriting this, highlighting the two lines that moved out of the function get moved to be inline with the opening bracket to `yada/listener` and sometimes running it a third time will kick everything out of `resource-route`.

I can reproduce the above issues consistently except for that very last bit about getting kicked out of the `resource-route`. The end result is the following code. 

```
(defn resource-routes []
    (yada/listener)
    ["/" (yada/handler (new File "app.html"))]
    {:port 5000})
```
that no matter if I try to manually move back into place produces "incorrect indentation for parinfer", actually further testing shows that this is not always the case but ends up back in a cyclical issues when running Code Reformat command. 
An aside this is the most used IntelliJ command I use and prior to installing the Parinfer (smart-mode) EAP build never caused me major issues that I couldn't manually fix.

Before
<img width="494" alt="screen shot 2018-03-19 at 22 39 24" src="https://user-images.githubusercontent.com/4560707/37626320-74c342b2-2bc7-11e8-8ac9-826798be61cf.png">
First Reformat
<img width="598" alt="screen shot 2018-03-19 at 22 39 36" src="https://user-images.githubusercontent.com/4560707/37626326-7dd13738-2bc7-11e8-8428-ccdee5482a1d.png">
Second Reformat
<img width="619" alt="screen shot 2018-03-19 at 22 39 46" src="https://user-images.githubusercontent.com/4560707/37626345-907cb682-2bc7-11e8-85f7-be7a4236dbf8.png">

**Comment by Severed-Infinity** [comment]
Further testing and it would seem its one of two things I've found that Code Reformat with Parinfer (smart-mode) don't like. They are as follows

- function arguments on a separate line, will close function call to close brackets orphaning off its arguments
- any form of list or collect (relates to the above) will maintain structure so long as everything is on the same line
    - if any item in a collection is not on the same line as the first item in the collection then they get orphaned (kicked out) and left to be lined up as a separate value/function/etc. on a new line potentially causing indentation errors.
- related to point two, trying to manually indent to correct location will result in "incorrect indentation for parinfer". 

These points make it feel as though I am back in the days of notepad with no formatting tools except space&tab at the moment.

**Comment by cursive-ide** [comment]
So I couldn't reproduce this until I realised that pasting your code in exactly did reproduce the problem. The issue seems to be that you're using 4 space indents when indenting your code manually, at least in the example above. For some reason which I'll investigate, reformatting that produces a different result when it's initially indented like this - if you use 2 space indents everything works fine. I still don't see exactly what you're seeing though:

If the code is indented with 2 space indents:

![screen shot 2018-03-20 at 12 54 42](https://user-images.githubusercontent.com/5589171/37628338-fbe34e00-2bd0-11e8-89be-1fdc510e29ae.png)

If the code is initially indented with 4 space indents, then it breaks on reformat:

![screen shot 2018-03-20 at 12 56 08](https://user-images.githubusercontent.com/5589171/37628377-25283bf4-2bd1-11e8-9c8c-74360da017ec.png)

However that still doesn't look quite like what you're seeing after your first reformat, though.

If you really do want 4-space indents, you can configure that in *Settings->Editor->Code Style->Clojure->Tabs and Indents*, and set both Tab size and Indent to 4. However I wouldn't recommend that, since 2 spaces is the style pretty much universally used for Clojure code.

Having played around with it a bit more, it's definitely an interaction with parinfer. If I turn structural editing off altogether, then reformatting works fine and corrects everything to two space indents. So if you do want to move your code to using 2 spaces, the best way would be to turn structural editing off, reformat your code and then turn parinfer back on.

**Comment by Severed-Infinity** [comment]
Yeah it was the indent size being the issue, I had to reinstall cursive and IntelliJ today as I was having lots of issues which got resolved by doing so but my preferences must have gotten deleted in the process. 

It maybe a minor issue with indent size and I understand norm of using 2 spaces (which I’ve always used) but for those rare instances for those who use large indent size it can be code breaking.
